### PR TITLE
Slack notifications

### PR DESCRIPTION
**Pull request by macat**
- adding new notification package
- triggering notifications after creation and deployment

### CODE REVIEW

File: services/helper_test.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+	"github.com/namely/broadway/env"
+)
+
+type notificationTestHelper struct {
+	requestBody string
+	ts          *httptest.Server
+}
+
+func newNotificationtestHelper() *notificationTestHelper {
```
Comment (poopoothegorilla) [inline services/helper_test.go]: not to be too picky but can we change this `newNotificationtestHelper()` to `newNotificationTestHelper()`

File: services/deployment_test.go [hunk 2]
```diff
@@ -11,6 +11,8 @@ import (
 )
 
 func TestDeployment(t *testing.T) {
+	nt := newNotificationtestHelper()
```
Comment (poopoothegorilla) [inline services/deployment_test.go]: here too just to follow the other naming patterns

**Comment by poopoothegorilla** [comment]
+1

File: env/env.go [hunk 3]
```diff
@@ -27,20 +27,24 @@ var (
...
+	ServerHost = loadw("HOST")
+	K8sServiceHost = loadw("KUBERNETES_SERVICE_HOST")
+	K8sServicePort = loadw("KUBERNETES_PORT_443_TCP_PORT")
+	K8sNamespace = loadf("KUBERNETES_NAMESPACE")
+	EtcdHost = loadw("ETCD_HOST")
+	EtcdPath = loadw("ETCD_PATH")
 }
 
-func load(key string) string {
+func loadw(key string) string {
```
Comment (cored) [inline env/env.go]: @macat why the change of name for this function?

File: notification/slack.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	"bytes"
+	"encoding/json"
+	"net/http"
+
+	"github.com/golang/glog"
+	"github.com/namely/broadway/env"
+)
+
+// Field is a Slack message field
+type Field struct {
```
Comment (cored) [inline notification/slack.go]: why just don't use the `SlackMessage` name then?

File: notification/slack.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	Title string `json:"title"`
+	Value string `json:"value"`
+	Short bool   `json:"short"`
+}
+
+// Attachment is Slack message attachment
+type Attachment struct {
```
Comment (cored) [inline notification/slack.go]: Why do we need to think about attachment ?

File: services/instance.go [hunk 6]
```diff
@@ -17,7 +20,17 @@ func NewInstanceService(s store.Store) *InstanceService {
 
 // Create a new instance
 func (is *InstanceService) Create(i *instance.Instance) error {
-	return is.repo.Save(i)
+	err := is.repo.Save(i)
+	if err != nil {
+		return err
+	}
+
+	err = sendCreationNotification(i)
```
Comment (cored) [inline services/instance.go]: I was thinking in more of the lines of: 
```
  instanceService := NewInstanceService(store, notifier)  // Notifier should implement the notification interface with the method Send() 
```
In that way we are actually exposing the relationship between the service and the notifications. And taking into account that you mention 'mail notification' at some point that approach will make more sense because the implementation detail will be interchangeable. 

/cc @bobbytables

File: notification/slack.go (hunk 5 above)
Comment (therealplato) [inline notification/slack.go]: for e.g. a red/black/green bar highlighting the bot response

File: services/instance.go (hunk 6 above)
Comment (therealplato) [inline services/instance.go]: i'd prefer `is.setNotifier()` otherwise it breaks the existing API, and does it really make sense to force the user to have a Notifier if they want to use InstanceService?

File: services/helper_test.go (hunk 1 above)
Comment (macat) [inline services/helper_test.go]: fixed

File: env/env.go (hunk 3 above)
Comment (macat) [inline env/env.go]: because there are now `loadf` and `loadw` methods

File: services/instance.go (hunk 6 above)
Comment (macat) [inline services/instance.go]: I agree with creating an abstraction, but only when we actually have more notifiers.

File: notification/slack.go (hunk 5 above)
Comment (macat) [inline notification/slack.go]: https://api.slack.com/incoming-webhooks

**Comment by therealplato** [comment]
+1

File: env/env.go (hunk 3 above)
Comment (cored) [inline env/env.go]: @macat right but why two different methods to load environment variables?

File: services/instance.go (hunk 6 above)
Comment (cored) [inline services/instance.go]: @macat at taht point we are going to miss the higher level interface for the notifiers and the refactor will be bigger. But ok.

File: env/env.go (hunk 3 above)
Comment (macat) [inline env/env.go]: `loadw` shows a warning, `loadf` exits if the env var is missing
